gcc/rust/backend/rust-compile-expr.cc

```diff
@@ -247,16 +252,30 @@ CompileExpr::visit (HIR::MatchExpr &expr)
   tree match_scrutinee_expr
     = CompileExpr::Compile (expr.get_scrutinee_expr ().get (), ctx);
 
-  // need to access the qualifier field, if we use QUAL_UNION_TYPE this would be
-  // DECL_QUALIFIER i think. For now this will just access the first record
-  // field and its respective qualifier because it will always be set because
-  // this is all a big special union
-  tree scrutinee_first_record_expr
-    = ctx->get_backend ()->struct_field_expression (
-      match_scrutinee_expr, 0, expr.get_scrutinee_expr ()->get_locus ());
-  tree match_scrutinee_expr_qualifier_expr
-    = ctx->get_backend ()->struct_field_expression (
-      scrutinee_first_record_expr, 0, expr.get_scrutinee_expr ()->get_locus ());
+  tree match_scrutinee_expr_qualifier_expr;
+  if (scrutinee_kind == TyTy::TypeKind::BOOL)
+    {
+      match_scrutinee_expr_qualifier_expr = match_scrutinee_expr;
+    }
+  else if (scrutinee_kind == TyTy::TypeKind::ADT)
+    {
+      // need to access qualifier the field, if we use QUAL_UNION_TYPE this
+      // would be DECL_QUALIFIER i think. For now this will just access the
+      // first record field and its respective qualifier because it will always
+      // be set because this is all a big special union
+      tree scrutinee_first_record_expr
+	= ctx->get_backend ()->struct_field_expression (
+	  match_scrutinee_expr, 0, expr.get_scrutinee_expr ()->get_locus ());
+      match_scrutinee_expr_qualifier_expr
+	= ctx->get_backend ()->struct_field_expression (
+	  scrutinee_first_record_expr, 0,
+	  expr.get_scrutinee_expr ()->get_locus ());
+    }
+  else
+    {
+      // FIXME: match on other types of expressions not yet implemented.
+      gcc_assert (0);
```

Comment:
```suggestion
      gcc_unreachable ();
```
nit but this way it's easier to grep for all those cases :)

Reply:
oh whoops I probably was thinking in `rust_assert (0)`.  `gcc_unreachable ()` is definitely better, but for some reason I wasn't thinking of using `gcc_x` for errors in the frontend. But of course it works :smile: